tests/func/test_odb.py

```diff
@@ -41,144 +41,144 @@ def test_get(self):
-        self.create(fname, "<clearly>not,json")
-        with pytest.raises(ObjectFormatError):
-            load(self.dvc.odb.local, HashInfo("md5", dir_hash))
+    dir_hash = "234.dir"
+    fname = os.fspath(dvc.odb.local.hash_to_path(dir_hash))
+    tmp_dir.gen({fname: '{"a": "b"}'})
+    with pytest.raises(ObjectFormatError):
+        load(dvc.odb.local, HashInfo("md5", dir_hash))
 
-        dir_hash = "234.dir"
-        fname = os.fspath(self.dvc.odb.local.hash_to_path(dir_hash))
-        self.create(fname, '{"a": "b"}')
-        with pytest.raises(ObjectFormatError):
-            load(self.dvc.odb.local, HashInfo("md5", dir_hash))
 
+def test_external_cache_dir(tmp_dir, dvc, make_tmp_dir):
+    cache_dir = make_tmp_dir("cache")
 
-class TestExternalCacheDir(TestDvc):
-    def test(self):
-        cache_dir = TestDvc.mkdtemp()
+    ret = main(["config", "cache.dir", cache_dir.fs_path])
+    assert ret == 0
 
-        ret = main(["config", "cache.dir", cache_dir])
-        self.assertEqual(ret, 0)
+    assert not os.path.exists(dvc.odb.local.cache_dir)
 
-        self.assertFalse(os.path.exists(self.dvc.odb.local.cache_dir))
+    tmp_dir.gen({"foo": "foo"})
+    ret = main(["add", "foo"])
+    assert ret == 0
 
-        ret = main(["add", self.FOO])
-        self.assertEqual(ret, 0)
+    tmp_dir.gen(
+        {
+            "data_dir": {
+                "data": "data_dir/data",
+                "data_sub_dir": {"data_sub": "data_dir/data_sub_dir/data_sub"},
+            }
+        }
+    )
+    ret = main(["add", "data_dir"])
+    assert ret == 0
 
-        ret = main(["add", self.DATA_DIR])
-        self.assertEqual(ret, 0)
+    assert not os.path.exists(".dvc/cache")
+    assert len(os.listdir(cache_dir)) != 0
 
-        self.assertFalse(os.path.exists(".dvc/cache"))
-        self.assertNotEqual(len(os.listdir(cache_dir)), 0)
 
-    def test_remote_references(self):
-        ssh_url = "ssh://user@localhost:23"
-        assert main(["remote", "add", "storage", ssh_url]) == 0
-        assert main(["remote", "add", "cache", "remote://storage/tmp"]) == 0
-        assert main(["config", "cache.ssh", "cache"]) == 0
+def test_remote_references(tmp_dir, dvc):
+    ssh_url = "ssh://user@localhost:23"
+    assert main(["remote", "add", "storage", ssh_url]) == 0
+    assert main(["remote", "add", "cache", "remote://storage/tmp"]) == 0
+    assert main(["config", "cache.ssh", "cache"]) == 0
 
-        self.dvc.__init__()
+    dvc.__init__()
 
-        assert self.dvc.odb.ssh.fs_path == "/tmp"
+    assert dvc.odb.ssh.fs_path == "/tmp"
 
 
-class TestSharedCacheDir(TestDir):
-    def test(self):
-        cache_dir = os.path.abspath(os.path.join(os.curdir, "cache"))
-        for d in ["dir1", "dir2"]:
-            os.mkdir(d)
-            os.chdir(d)
+def test_shared_cache_dir(tmp_dir, scm, dvc):
+    cache_dir = os.path.abspath(os.path.join(os.curdir, "cache"))
+    for d in ["dir1", "dir2"]:
+        os.mkdir(d)
+        os.chdir(d)
```

Comment:
This test should really be re-written to use the `tmp_dir` fixture properly. It also does not need the `scm` or `dvc` fixtures, since the test itself is initializing DVC repos with `--no-scm`.

So the test should be something like:
```py
def test_shared_cache_dir(tmp_dir):
    cache_dir = tmp_dir / "cache"
    for d in ["dir1", "dir2"]:
        makedirs(tmp_dir / d)
        with (tmp_dir / d).chdir():
            # do the rest of the test operations
            ...
    assert not (tmp_dir / "dir1" /  ".dvc" / "cache").exists()
    assert not (tmp_dir / "dir2" /  ".dvc" / "cache").exists()
    ...
```

We can also just use `(tmp_dir / d).gen(...)` instead of the existing `with open(...): write(...)` calls as well